### ISSUE TITLE
Bootstrap 4 support

### ISSUE THREAD

**Issue by jeevcat** [mbr/flask-bootstrap#162]
Is there an easy way to switch over to Bootstrap 4, while using this extension? I understand it's still in Alpha, but I figured it might be something other people might also be interested in.

**Comment by mbr** [comment]
Unfortunately, I am a bit strapped for time and haven't had a look at
Bootstrap 4 yet - I'm waiting for it to be released first. Your best option
is a fork to try it out.

Usually changing version numbers is straightforward, you'll have to find
out yourself how backwards-incompatible the whole thing really is =).

**Comment by jeevcat** [comment]
Okay thanks for the reply. 

I've forked the repo and managed to get v4.0.0-Alpha.6 working. I had to include Tether.js which is required by Bootstrap 4 for tooltips apparently. I also had to fiddle a little with the versioning regex due to the -alpha-style version numbers.

**Comment by cofinley** [comment]
@jeevcat If you have something that works, I'd love to check it out on your fork once you push it.

**Comment by jeevcat** [comment]
I actually already pushed it 2 days ago. Check it out here: https://github.com/jeevcat/flask-bootstrap.

Please let me know if it works for you.

**Comment by cofinley** [comment]
I was mainly looking for updated wtforms. This is a step in the right direction though. I've updated the forms part on my clone, but kind of towards a side project which may not fall in line with the styling here. Thank you again.

**Comment by OverlordQ** [comment]
Yeah, the main issue I've found is the WTForms macros will need a heavy overhaul for v4 forms.

**Comment by beenje** [comment]
I'll try to make a proper pull request starting from @jeevcat repository.

**Comment by beenje** [comment]
I just created the pull request https://github.com/mbr/flask-bootstrap/pull/164

**Comment by MelleB** [comment]
What do we need to push this forward?

**Comment by mbr** [comment]
I will be looking into this on Sunday.

The current blocker is that I want to deprecate quick_forms in favor of the
new, dominate-based quick_forms which has *much* more maintaineable code.
Otherwise, I would have done this already.

**Comment by mbr** [comment]
And FWIW, I am sorry for the delay =(

**Comment by cofinley** [comment]
Recent bug I found with the 4.0 branch is that pagination.html needs the appropriate .page-link and .page-item classes.

**Comment by mcnullity** [comment]
Eagerly waiting...

**Comment by joncutrer** [comment]
+1

**Comment by slidenerd** [comment]
Bootstrap 4 has moved well into the beta version guys, any updates on this?

**Comment by zgoda** [comment]
B4 left beta back in January, now it's at 4.1.1.

**Comment by greyli** [comment]
For people who interested, I am maintaining [Bootstrap-Flask](https://github.com/greyli/bootstrap-flask) that based on Flask-Bootstrap and **support Bootstrap 4**.

**Comment by louisguitton** [comment]
Would love to see this as well, as I'm using cards for example.

**Comment by jaikumarm** [comment]
+1

**Comment by joaqo** [comment]
+1

**Comment by utkonos** [comment]
See my answer in https://github.com/mbr/flask-bootstrap/issues/192
I was able to update all versions to current just editing `__init__.py` in site-packages with vi.

**Comment by hfilimonescu** [comment]
Hi Marc, 
I was wondering why all this trouble with dominate? WTForms generate the html with __call__. Am I missing something? 

I ask because I might be prepared to do some work on this issue, but first I need to understand the choices and restrictions. 

Best regards

**Comment by yurzo** [comment]
Same here, I'm new to Bootstrap and Flask yet the docs and ideas behind flask-bootstrap are gorgeous so I'm interested in seeing it move forward

**Comment by bartkamphorst** [comment]
Ping @mbr. Would also love for bootstrap 4 to be supported. Thanks in advance!

**Comment by stratisastro** [comment]
Likewise ping @mbr. Really looking to move to bootstrap 4 (including with WTF) so be good to know if this is imminent or whether I should roll my own. TIA!

**Comment by 0x78f1935** [comment]
I noticed a lot of people want to use bootstrap4. If you made yourself a base template and if you are using flask-bootstrap then this might work for you too.

```
<!doctype html>
<html lang="en">
    <head>
        {% block title %}
        <title>{% if title %}{{ title }}{% else %}My Website{% endif %}</title>
        {% endblock title %}
        
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        
        {% block meta %}
        {% endblock meta %}

        {% block import %}
        {% endblock import %}

        <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css" integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">

        {% block style %}
        {% endblock style %}
    </head>
    <body>
        {% block navbar %}
        {% endblock navbar %}

        <div id='d-flex p-2 bd-highlight'>
            {% block content %}
            {% endblock content %}
        < div >

        {% block scripts %}
        { % endblockscripts % }
        
        <script src="https://code.jquery.com/jquery-3.4.1.js" integrity="sha256-WpOohJOqMqqyKL9FccASB9O0KwACQJpFTUBLTYOVvVU=" crossorigin="anonymous"></script>
        <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js" integrity="sha384-JjSmVgyd0p3pXB1rRibZUAYoIIy6OrQ6VrjIEaFf/nJGzIxFDsf4x0xIM+B07jRM" crossorigin="anonymous"></script>
        
    </body>
</html>
```
I do use this myself and it works fine in my situation. I hope this helps out.

**Comment by bartkamphorst** [comment]
@mbr Any updates regarding bootstrap 4 support?

**Comment by hfilimonescu** [comment]
@bartkamphorst I have made a new package witg Bootstrap 4, based on Flask-Bootstrap.
If you want, you can give it a spin. [Flask-BS4](https://pypi.org/project/Flask-BS4/)

**Comment by maxtrem** [comment]
This issue has been open now for 3 years. Seems to be quite dead. 

> on 10 Apr 2018:
> I will be looking into this on Sunday.


Will there still be a version for bootstrap 4?

**Comment by mpounsett** [comment]
@mbr Maybe it's time to find a new maintainer for this project?
